logstash-rs/src/buffer.rs

```diff
@@ -0,0 +1,152 @@
+use crate::prelude::*;
+use std::{
+    sync::mpsc,
+    time::{Duration, Instant},
+};
+
+enum Command {
+    Send(LogStashRecord),
+    SendBatch(Vec<LogStashRecord>),
+    Flush,
+}
+
+pub struct BufferedSender {
+    sender: mpsc::SyncSender<Command>,
+}
+
+impl BufferedSender {
+    pub fn new<S: Sender>(
+        sender: S,
+        buffer_size: Option<usize>,
+        buffer_lifetime: Option<Duration>,
+    ) -> Self {
+        let sender = BufferedSenderThread::new(sender, buffer_size, buffer_lifetime).run();
+        Self { sender }
+    }
+}
+
+impl Sender for BufferedSender {
+    fn send(&self, event: LogStashRecord) -> Result<()> {
+        self.sender
+            .send(Command::Send(event))
+            .map_err(|_| anyhow::anyhow!("Failed to send record to channel"))?;
```

Comment:
Probably it's better to add inner error's message to the new error, so it's easier to find problems

Reply:
Done. Now error type implemented with `thiserror` crate

Reply:
LGTM. But you probably can add `#[source]` or `#[from]` attributes to Error's fields, so original error is preserved.